### PR TITLE
LDpred2.R uses external ld matrix (partial fix for #108)

### PR DESCRIPTION
**Pull request by ofrei**
Fixes first half of #108

Changes proposed in this pull request:

- remove LD matrix computation from LDpred2.R script
- instead, load reference data submitted to https://github.com/comorment/ldpred2_ref repository. For now the path to the reference was hard-coded (fell free to fix, or I can fix later)
- allow to read subset of chromosomes using ``--chr2use`` flag
- overlap the set of SNPs from (1) sumstats file (2) genotype (3) LD reference. Use [snp_match](https://privefl.github.io/bigsnpr/reference/snp_match.html) function to perform this logic. Prefer merging on CHR:BP:A1:A2, give user a new ``--merge-by-rsid`` option to perform merging on rs# (SNP marker name). Test ``snp_match`` with a new unit-test, ``tests/extras/bigsnpr.R`` (the test passes)
- implement new dataset (see ``reference/examples/ldpred2`` folder), which allows to test how well ``LDpred2`` matches SNPs from sumstats, genotypes and LD reference. The tool also runs very fast on that dataset because it only includes chr21 and chr22.
- remove run_ldpred2.sh; usage example integrated into README.md

## Before submitting

- [x] I've read and followed all steps in the [Making a pull request](https://github.com/comorment/containers/blob/main/CONTRIBUTING.md#making-a-pull-request)
    section of the `CONTRIBUTING` docs.
- [x] I've updated or added any relevant docstrings following the syntax described in the
    [Writing docstrings](https://github.com/comorment/containers/blob/main/CONTRIBUTING.md#writing-docstrings) section of the `CONTRIBUTING` docs.
- [ ] If this PR fixes a bug, I've added a test that will fail without my fix.
- [x] If this PR adds a new feature, I've added tests that sufficiently cover my new functionality.

### CODE REVIEW

**Comment by ofrei** [comment]
I'll fix Flake8 linter errors later today. The LD reference files are stil being uploaded to https://github.com/comorment/ldpred2_ref . Apart from this this PR is ready for review.

**Comment by ofrei** [comment]
@deepchocolate @espenhgn I pushed few more changes, that's would be all I wanted to implement for ldpred2.R. Is it fine to merge? I'd like to ask our colleagues at CoMorMent to have a look before the meeting tomorrow.

@deepchocolate what do you think about adding --geno-file, --sumstats  and --out as named arguments, rather than using them as positional arguments? I generally prefer named arguments over positional arguments, also for those arguments that are required. To me this is more readable and intuitive as one doesn't need to know that order in which positional arguments must be provided. Also it's fine to reorder without breaking backwards compatibility. Finally, with --chr2use argument with ``narg=Inf`` the behaviour is non-intuitive if ``--chr2use`` is the last argument before positional arguments, e.g. ``... --chr2use 21 22 $genoFile $sumstatsFile $outFile``. In this situation  R parser doesn't have a way to know that there are only 2 arguments for --chr2use, the other three are positional. The command fails but  that's really hard for the user to investigate.

**Comment by deepchocolate** [comment]
> @deepchocolate @espenhgn I pushed few more changes, that's would be all I wanted to implement for ldpred2.R. Is it fine to merge? I'd like to ask our colleagues at CoMorMent to have a look before the meeting tomorrow.
> 

I'm looking at it now, I'll approve today, just checking if I need to do some changes somewhere. If there are any I'll put them in a separate branch

> @deepchocolate what do you think about adding --geno-file, --sumstats and --out as named arguments, rather than using them as positional arguments? I generally prefer named arguments over positional arguments, also for those arguments that are required. To me this is more readable and intuitive as one doesn't need to know that order in which positional arguments must be provided. Also it's fine to reorder without breaking backwards compatibility. Finally, with --chr2use argument with `narg=Inf` the behaviour is non-intuitive if `--chr2use` is the last argument before positional arguments, e.g. `... --chr2use 21 22 $genoFile $sumstatsFile $outFile`. In this situation R parser doesn't have a way to know that there are only 2 arguments for --chr2use, the other three are positional. The command fails but that's really hard for the user to investigate.

That's fine by me, good point!  Also good to harmonize arguments to gwas.py, didn't think about that as I usually try to prefix the arguments after what they specify (`--file-...`).

**Comment by ofrei** [comment]
@deepchocolate I agree about ``--file-``, this sounds good idea to me, and currently this is also inconsistent in gwas.py (e.g. --sumstats also points to the file same as --geno-file, but for no good reason only one of them has ``-file``). I suggest for now we keep things consistent with ``gwas.py``, discuss the ideal names when there is more time, and change naming for in a later release.

**Review by espenhgn — CHANGES_REQUESTED** [review]

File: usecases/LDpred2/run_ldpred2_slurm.job [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+export fileOut=Height
+export colPheno=Height
+
+# set environmental variables. Replace $COMORMENT with 
+# the full path to the folder containing cloned "containers" and "ldpred2_ref" repositories
+export COMORMENT=/ess/p697/data/durable/s3-api/github/comorment
+export SIF=$COMORMENT/containers/singularity
+export REFERENCE=$COMORMENT/containers/reference
+export LDPRED2_REF=$COMORMENT/ldpred2_ref
+export SINGULARITY_BIND=$REFERENCE:/REF,${LDPRED2_REF}:/ldpred2_ref
```
[the remark this answers is not in the document]
Comment (ofrei) [reply usecases/LDpred2/run_ldpred2_slurm.job:30]: Yes, that's correct. I suggest all comorment repos (http://github.com/comorment/containers, https://github.com/comorment/ldpred2_ref etc) are cloned to one root-level folder  which is referred to as $COMORMENT.

File: usecases/LDpred2/README.md [hunk 2]
```diff
@@ -47,47 +75,39 @@ HG00100 HG00100 NA -1.8332542097235e+100
 ...
 ```
 
-The ``run_ldpred2.sh`` script will also output ``.bk`` and ``.rds`` binary files with prefix ``EUR`` in this directory.
+The script will also output ``.bk`` and ``.rds`` binary files with prefix ``EUR`` in this directory.
 
 
-## Slurm job
+## Running LDpred2 analysis - synthetic example (chr21 and chr22)
```
[the remark this answers is not in the document]
Comment (ofrei) [reply usecases/LDpred2/README.md:81]: These files together take around 250 MB, perhaps too big for a demo. I suggest we keep them in ``ldpred2_ref`` repo for now, and later we can generate our own small-scale demo LD example using ``calculateSnpCorrelation.R`` script that @deepchocolate will develop as part of #108

[the remark this answers is not in the document]
Comment (espenhgn) [reply usecases/LDpred2/README.md:81]: I should add: I could run both the infinitesimal and auto models without issues (chr 21-22). At least the p/h2 vs. iteration plots now make sense.

File: usecases/LDpred2/run_ldpred2_slurm.job (hunk 1 above)
[the remark this answers is not in the document]
Comment (deepchocolate) [reply usecases/LDpred2/run_ldpred2_slurm.job:30]: Ah, I see!

File: usecases/LDpred2/ldpred2.R [hunk 3]
```diff
@@ -5,65 +5,69 @@ options(bigstatsr.check.parallel.blas = FALSE)
 options(default.nproc.blas = NULL)
 library(tools)
 library(argparser, quietly=T)
+library(stringr)
```
[the remark this answers is not in the document]
Comment (ofrei) [reply usecases/LDpred2/ldpred2.R:8]: :+1: will fix

File: usecases/LDpred2/ldpred2_simulations.ipynb [hunk 4]
```diff
@@ -0,0 +1,531 @@
...
+  {
+   "cell_type": "code",
+   "execution_count": 27,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "import pandas as pd\n",
+    "import numpy as np\n",
+    "\n",
+    "bfile = '/home/oleksanf/github/comorment/magma/reference/magma/g1000_eur/g1000_eur'\n",
```
[the remark this answers is not in the document]
Comment (ofrei) [reply usecases/LDpred2/ldpred2_simulations.ipynb:20]: Added brief description on top, that's for sure a good idea.

[the remark this answers is not in the document]
Comment (ofrei) [reply usecases/LDpred2/ldpred2_simulations.ipynb:20]: Reproducing from within container is also good, but I haven't tested it yet.

File: usecases/LDpred2/README.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (ofrei) [reply usecases/LDpred2/README.md]: Great! The examples are swapped now, so chr21-22 is described first, followed by full example

**Comment by ofrei** [comment]
I rebuilt r.sif container including ``stringr`` library, and fix other thing we discussed with @espenhgn 
Also there is now a note on handling missing genotypes: https://github.com/comorment/containers/tree/of_ldpred2/usecases/LDpred2#handling-missing-genotypes
OK to merge?

**Review by deepchocolate — APPROVED** [review]
Great!

**Pull request merged**